pkg/controller/certificate-shim/sync.go

```diff
@@ -140,35 +159,105 @@ func validateIngressTLSBlock(tlsBlock networkingv1beta1.IngressTLS) []error {
 	return errs
 }
 
-func (c *controller) buildCertificates(ctx context.Context, ing *networkingv1beta1.Ingress,
+func validateGatewayListeners(listeners []gatewayapi.Listener) []error {
+	var errs []error
+
+	for _, l := range listeners {
+		if l.TLS == nil {
+			return []error{fmt.Errorf("listener for host %s does not specify a TLS block", *l.Hostname)}
+		}
+		if l.TLS.CertificateRef.Group != "core" {
+			errs = append(
+				errs,
+				fmt.Errorf(
+					"unsupported certificateRef.group, want %s got %s",
+					"core",
+					l.TLS.CertificateRef.Group,
+				),
+			)
+		}
+		if *l.TLS.Mode == gatewayapi.TLSModePassthrough {
+			errs = append(
+				errs,
+				fmt.Errorf(
+					"listener for %s has mode %s, the only valid mode is %s",
+					*l.Hostname,
+					gatewayapi.TLSModePassthrough,
+					gatewayapi.TLSModeTerminate,
+				),
+			)
+		}
+	}
+	return errs
+}
+
+func (c *controller) buildCertificates(ctx context.Context, obj metav1.Object,
 	issuerName, issuerKind, issuerGroup string) (new, update []*cmapi.Certificate, _ error) {
 	log := logf.FromContext(ctx)
 
 	var newCrts []*cmapi.Certificate
 	var updateCrts []*cmapi.Certificate
-	for i, tls := range ing.Spec.TLS {
-		errs := validateIngressTLSBlock(tls)
-		// if this tls entry is invalid, record an error event on Ingress object and continue to the next tls entry
+
+	type certificateShimInfo struct {
+		// tlsHosts key = secret ref, value = dns host names
+		tlsHosts map[corev1.ObjectReference][]string
+		gvk      schema.GroupVersionKind
+	}
+
+	certs := certificateShimInfo{
+		tlsHosts: map[corev1.ObjectReference][]string{},
+	}
+
+	switch o := obj.(type) {
+	case *networkingv1beta1.Ingress:
+		certs.gvk = ingressGVK
+		for i, tls := range o.Spec.TLS {
+			errs := validateIngressTLSBlock(tls)
+			// if this tls entry is invalid, record an error event on IngressOrGateway object and continue to the next tls entry
```

Comment:
nit: not sure whether this comment is helpful to the reader; the below code feels straightforward enough to be self-documenting